Implement matching of image tag patterns

broesch opened the pull request:
- Implemented matching of go-glob pattern for image tags.
- Added ryanuber/go-glob
- Modified tests to include “*”

Commit:
Implement matching of image tag patterns

- Implemented matching of go-glob pattern for image tags.
- Added ryanuber/go-glob
- Modified tests to include “*”

mrajashree commented:
Hi @broesch , can you split your PR into two commits, one for the code changes and other for the vendor folder changes?

broesch commented:
I was away for a couple of days, I'll get to it soon, stay tuned...

Commit:
Implement matching of image tag patterns

- Added ryanuber/go-glob

Commit:
Implement matching of image tag patterns

- Implemented matching of go-glob pattern for image tags.
- Modified tests to include “*”

Commit:
Merge branch 'matching-image-tag-patterns' of https://github.com/broesch/webhook-service into matching-image-tag-patterns

mrajashree commented on drivers/service_upgrade.go:
```diff
@@ -263,7 +264,7 @@ func wait(apiClient *client.RancherClient, service *client.Service) error {
 	match := regTag.FindAllString(tag, -1)
 	if len(match) == 0 || len(match[0]) > 128 || (len(match[0]) != len(tag)) {
-		return fmt.Errorf("Invalid tag %s, tag length must be < 128, must contain [a-zA-Z0-9.-_] characters only, cannot start with '.' or '-'", tag)
+		return fmt.Errorf("Invalid tag %s, tag length must be < 128, must contain [a-zA-Z0-9.-_*] characters only, cannot start with '.' or '-'", tag)
```
@broesch we wanted to use tags that are allowed by docker as per this document https://docs.docker.com/engine/reference/commandline/tag/#extended-description
In this `*` isn't a valid character.

WoLfulus replied:
@mrajashree @broesch how's this going? we were using webhooks to deploy our platform and we just found out that we can't do versioning of tags because of this :(

broesch commented on drivers/service_upgrade.go:
```diff
@@ -263,7 +264,7 @@ func wait(apiClient *client.RancherClient, service *client.Service) error {
 	match := regTag.FindAllString(tag, -1)
 	if len(match) == 0 || len(match[0]) > 128 || (len(match[0]) != len(tag)) {
-		return fmt.Errorf("Invalid tag %s, tag length must be < 128, must contain [a-zA-Z0-9.-_] characters only, cannot start with '.' or '-'", tag)
+		return fmt.Errorf("Invalid tag %s, tag length must be < 128, must contain [a-zA-Z0-9.-_*] characters only, cannot start with '.' or '-'", tag)
```
@mrajashree So which character do you suggest as wildcard? If we choose one that is valid in docker tags, we would have no way of knowing if the user wanted wildcard matching or exactly this tag. (Also * is wildly used as wildcard)

orangewolf commented:
I'd like to 👍 this.  having a wildcard here would be a big win

WoLfulus commented:
@orangewolf Yeah, we had to implement an external agent to receive webhooks and upgrade the services when tags were versioned.  :(

Commit:
Merge branch 'master' into matching-image-tag-patterns

Commit:
Merge remote-tracking branch 'rancher/master' into matching-image-tag-patterns

# Conflicts:
#	drivers/service_upgrade.go

Commit:
Merge remote-tracking branch 'origin/matching-image-tag-patterns' into matching-image-tag-patterns

# Conflicts:
#	drivers/service_upgrade.go

broesch commented:
I fixed the merge conflict (branch had diverged).
@mrajashree How do we proceed? Should something be refactored?

cjellick commented:
Very stale. Closing. Thanks! Please open a new PR if there is still a relevant change.

defcon8 commented:
We need this.... why is this still no part of Rancher? Not being able to use wildcards or regex in the tag makes the webhook / update service function totally useless :(
